models/stock_location.py

```diff
@@ -54,12 +71,113 @@ def get_location(self, location_identifier):
+        if PI_COUNT_MOVES in request:
+            for obj in request[PI_COUNT_MOVES]:
+                self._check_locations(['location_id', 'location_dest_id'], obj)
+
+        if PRECEDING_INVENTORY_ADJUSTMENTS in request:
+            if not request.get(INVENTORY_ADJUSTMENTS):
+                raise ValidationError(
+                    _('You must specify inventory adjustments if you require '
+                      'preceding adjustments.'))
+
+            self._check_request_locations(['location_id'],
+                                          request[INVENTORY_ADJUSTMENTS])
+
+    def process_perpetual_inventory(self, request):
+        """
+            Executes the specified PI request by processing PI count
+            moves, inventory adjustments, and preceding inventory
+            adjustments. Updates the PI date time attributes (last
+            check dates) accordingly.
+
+            Raises a ValidationError in case of invalid request (e.g.
+            one of the specified locations doesn't exist).
+
+            Returns True in case any change as been made, False
+            otherwise.
+        """
+        self._validate_perpetual_inventory_request(request)
+        pi_outcome = defaultdict()
+
+        if PI_COUNT_MOVES in request:
+            pi_outcome[PI_COUNT_MOVES] = \
+                self._process_pi_count_moves(request[PI_COUNT_MOVES])
+
+        if INVENTORY_ADJUSTMENTS in request:
+            adjs = request[INVENTORY_ADJUSTMENTS]
+            adj_inv = self._process_inventory_adjustments(adjs)
+
+            for pre_adj in request.get(PRECEDING_INVENTORY_ADJUSTMENTS, []):
+                self._process_preceding_inventory_adjustments(pre_adj, adj_inv)
+
+            pi_outcome[INVENTORY_ADJUSTMENTS] = adj_inv
+
+        self._process_pi_datetime(pi_outcome)
+
+        # @todo: reponse format?
+        return bool(pi_outcome)
+
+    def _process_pi_count_moves(self, moves_request):
+        """
+            Returns the modified inventory in case changes were
+            necessary, None otherwise.
+        """
+        # @todo
+        pass
+
+    def _process_inventory_adjustments(self, adjustments_request):
+        """
+            Returns the modified inventory in case changes were
+            necessary, None otherwise.
+        """
+        # @todo
+        pass
+
+    def _process_preceding_inventory_adjustments(self, pre_adj, next_adj_inv):
+        Location = self.env['stock.location']
+
+        # @todo: do we need casting here?
+        location_id = int(pre_adj['location_id'])
+        location = Location.browse(location_id)
+        other_adjs = pre_adj[INVENTORY_ADJUSTMENTS]
+        inv = location._process_inventory_adjustment(other_adjs)
+        inv.u_next_inventory_id = next_adj_inv
+
+    def _process_pi_datetime(self, pi_outcome):
+        current_time = datetime.now()
+        self.write({'u_date_last_checked': current_time})
+
+        if all([pi_outcome[key] is None
```

Comment:
If pi_outcome does not contain some of the keys is going to fail.

Reply:
This is equivalent to prior art.